src/mlpack/methods/edge_boxes/feature_extraction_impl.hpp

```diff
@@ -0,0 +1,1052 @@
+ */
+/*Implement this function in a column major order.*/
+template<typename MatType, typename CubeType>
+void StructuredForests<MatType, CubeType>::
+BilinearInterpolation(const MatType& src,
+                      size_t height, size_t width,
+                      MatType& dst)
+{
+  dst = MatType(height, width);
+  double const x_ratio = static_cast<double>((src.n_cols - 1)) / width;
+  double const y_ratio = static_cast<double>((src.n_rows - 1)) / height;
+  for(size_t row = 0; row != dst.n_rows; ++row)
+  {
+    size_t y = static_cast<size_t>(row * y_ratio);
+    double const y_diff = (row * y_ratio) - y; //distance of the nearest pixel(y axis)
+    double const y_diff_2 = 1 - y_diff;
+    for(size_t col = 0; col != dst.n_cols; ++col)
+    {
+      size_t x = static_cast<size_t>(col * x_ratio);
+      double const x_diff = (col * x_ratio) - x; //distance of the nearet pixel(x axis)
+      double const x_diff_2 = 1 - x_diff;
+      double const y2_cross_x2 = y_diff_2 * x_diff_2;
+      double const y2_cross_x = y_diff_2 * x_diff;
+      double const y_cross_x2 = y_diff * x_diff_2;
+      double const y_cross_x = y_diff * x_diff;
+      dst(row, col) = y2_cross_x2 * src(y, x) +
+              y2_cross_x * src(y, x + 1) +
+              y_cross_x2 * src(y + 1, x) +
+              y_cross_x * src(y + 1, x + 1);
+    }
+  }
+}
+
+/**
+ * Applies a separable linear filter to an Image
+ * @param InOutImage Input/Output Contains the input Image, The final filtered Image is
+ *          stored in this param.
+ * @param kernel Input Kernel vector to be applied on Image.
+ * @param radius amount, the Image should be padded before applying filter.
+ */
+template<typename MatType, typename CubeType>
+void StructuredForests<MatType, CubeType>::
+SepFilter2D(CubeType &InOutImage, const arma::vec& kernel, const size_t radius)
+{
+  CubeType OutImage;
+  this->CopyMakeBorder(InOutImage, radius, radius, radius, radius, OutImage);
+
+  arma::vec row_res, col_res;
+  // reverse InOutImage and OutImage to avoid making an extra matrix.
+  // InImage is renamed to InOutImage in this function for this reason only.  
+  MatType k_mat = kernel * kernel.t();
+  for(size_t k = 0; k < OutImage.n_slices; ++k)
+  {
+    for(size_t j = radius; j < OutImage.n_cols - radius; ++j)
+    {
+      for(size_t i = radius; i < OutImage.n_rows - radius; ++i)
+      {
+        InOutImage(i - radius, j - radius, k) = 
+            arma::accu(OutImage.slice(k)\
+            .submat(i - radius, j - radius,\
+              i + radius, j + radius) % k_mat);
+      }
+    }
+  }
+
+}
+
+/**
+ * Applies a triangle filter on an Image.
+ * @param InImage Input/Output Image on which filter is applied.
+ * @param radius Decides the size of kernel to be applied on Image.
+ */
+template<typename MatType, typename CubeType>
+void StructuredForests<MatType, CubeType>::
+ConvTriangle(CubeType &InImage, const size_t radius)
+{
+  if (radius == 0)
+  {
+    //nothing to do
+  }
```

Comment:
A little bit weird to see empty if, since radius is size_t, you can
```
if(radius != 0){
  if(radius == 1){
  //.....
  }else{
  //....
 }
}
```